File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # [1, 2, 3, 4]
        # do prefix[i] = prefix and then update prefix = prefix * nums[i]
        # pre = 24 product = [24, 12, 8, 6]
        # suffix= 24
        # going backwards product[i] = suffix*product[i], suffix *= nums[i] 
        prefix = 1
        suffix = 1
        product = [1] * len(nums)
        for i in range(len(nums)):
            product[i] = prefix
            prefix *= nums[i]
        for i in range(len(nums)-1, -1, -1):
            product[i] *= suffix
            suffix *= nums[i]
        return product

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answers = new int[nums.length]; // nums: [1, 2, 3, 4]
        int prefix = 1; // 24
        for (int i = 0; i < nums.length; i++) { // [1, 1, 2, 6]
            answers[i] = prefix;
            prefix *= nums[i];
        }
        int suffix = 1; // 12
        for (int i = nums.length-1; i >= 0; i--) { // [24 , 12 , 8 , 6 ]
            answers[i] *= suffix;
            suffix *= nums[i];
        }
        return answers;
    }
}